File: python/040.py
#   040
#   Combination Sum II
#   2015-05-16
#####################################################
class Solution:
    # @param {integer[]} candidates
    # @param {integer} target
    # @return {integer[][]}
    def combinationSum2(self, candidates, target):
        candidates.sort()
        return self.solve(candidates, target, len(candidates))

    def solve(self, candidates, target, limit):
        v = []
        if target < candidates[0]:
            return v
        for i in list(reversed(range(limit))):
            if (candidates[i] == target):
                v.append([target])
                limit = i
                break
        for i in list(reversed(range(limit))):
            if i < limit-1 and candidates[i] == candidates[i+1]:
                continue
            for a in self.solve(candidates, target-candidates[i], i):
                a.append(candidates[i])
                v.append(a)
        return v

#####################################################

candidates = [1,1,1,3,3,5]
target = 8

sol = Solution()
for a in sol.combinationSum2(candidates, target):
    print(a)

File: ruby/040.rb
# 040
# Combination Sum II
# 2015-05-16
#####################################################
# @param {Integer[]} candidates
# @param {Integer} target
# @return {Integer[][]}
def combination_sum2(candidates, target)
  candidates.sort!()
  solve(candidates, target, candidates.size)
end

def solve(candidates, target, limit)
  v = []
  return v if target < candidates.first
  (limit-1).downto(0) do |i|
    if candidates[i] == target
      v.push([target])
      limit = i
      break
    end
  end
  (limit-1).downto(0) do |i|
    next if i < limit-1 and candidates[i] == candidates[i+1]
    solve(candidates, target-candidates[i], i).each do |a|
      a.push(candidates[i])
      v.push(a)
    end
  end
  v
end

#####################################################

candidates = [1,1,1,3,3,5]
target = 8

candidates = [2,5,2,1,2]
target = 5

combination_sum2(candidates, target).each do |a|
  puts a.inspect
end

